Framework/Core/include/Framework/ContextRegistry.h

```diff
@@ -11,55 +11,63 @@
 #ifndef FRAMEWORK_CONTEXTREGISTRY_H
 #define FRAMEWORK_CONTEXTREGISTRY_H
 
-#include <array>
+#include <unordered_map>
+#include <typeinfo>
+#include <typeindex>
+#include <type_traits>
+#include <string>
+#include <stdexcept>
 
 namespace o2
 {
 namespace framework
 {
 
+/// @class ContextRegistry
+/// Instances are registered by pointer and are not owned by the registry
 /// Decouples getting the various contextes from the actual type
 /// of context, so that the DataAllocator does not need to know
-/// about the various serialization methods. Since there is only
-/// a few context types we support, this can be done in an ad hoc
-/// manner making sure each overload of ContextRegistry<T>::get()
-/// uses a different entry in ContextRegistry::contextes;
+/// about the various serialization methods.
 ///
-/// Right now we use:
-///
-/// MessageContext 0
-/// ROOTObjectContext 1
-/// StringContext 2
-/// ArrowContext 3
-/// RawContext 4
 class ContextRegistry
 {
  public:
-  ContextRegistry(std::array<void*, 5> contextes)
-    : mContextes{ contextes }
+  ContextRegistry();
+
+  template <typename... Types>
+  ContextRegistry(Types*... instances)
+  {
+    set(std::forward<Types*>(instances)...);
```

Comment:
I guess you can use something along the lines of:

```C++
(set(std::forward<Types*>(instances)), ...)
```

and have `set` work on a single entry.

Reply:
need a function which one can iterate on,if not calling the constructor multiple times. Having set with argument pack seems natural to me. In addition, to use set with an arbitrary list of arguments is also possible, and the actual sequence of types does not matter.

Reply:
My point is that you do not need the function, but you can use the comma operator.

```C++
(somefunction(something), ...)
```

is actually a valid way of specifying a fold expression. See:

https://en.cppreference.com/w/cpp/language/fold

I use it in:

https://github.com/AliceO2Group/AliceO2/blob/7e99dde7bbe9f22d788ef99a71ef6f7dc05bcecc/Framework/Core/include/Framework/TableBuilder.h#L201